src/main/clojure/clara/rules/engine.cljc

```diff
@@ -1748,50 +1748,74 @@
         (when (flush-updates *current-session*)
           (recur (mem/next-activation-group transient-memory) next-group))))))
 
-(deftype LocalSession [rulebase memory transport listener get-alphas-fn]
+(deftype LocalSession [rulebase memory transport listener get-alphas-fn pending-operations]
   ISession
   (insert [session facts]
-    (let [transient-memory (mem/to-transient memory)
-          transient-listener (l/to-transient listener)]
 
-      (l/insert-facts! transient-listener facts)
-
-      (binding [*pending-external-retractions* (atom [])]
-        ;; Bind the external retractions cache so that any logical retractions as a result
-        ;; of these insertions can be cached and executed as a batch instead of eagerly realizing
-        ;; them.  An external insertion of a fact that matches
-        ;; a negation or accumulator condition can cause logical retractions.
-        (doseq [[alpha-roots fact-group] (get-alphas-fn facts)
-                root alpha-roots]
-          (alpha-activate root fact-group transient-memory transport transient-listener))
-        (external-retract-loop get-alphas-fn transient-memory transport transient-listener))
+    (let [new-pending-operations (conj pending-operations (->PendingUpdate :insertion (if (coll? facts)
+                                                                                        facts
+                                                                                        (into [] facts))))]
```

Comment:
What is the purpose of putting the `facts` into a vector here?  It seems like it could just end up being a waste of time.  Are you concerned with the mutability of the collection?  I'm not sure if that is a concern that should be paid for with a perf penalty like this.  If the caller is mutating the same collection of `facts` they passed to the session, the expected semantics would be mysterious anyways.

I'd say do `seq` instead, but it'd still be affected by the backing collection beyond the first (32 element) chunk.